contracts/abstracts/TokenCollector.sol

```diff
@@ -70,8 +74,25 @@ abstract contract TokenCollector {
                 data,
                 (address, IUniswapPermit2.PermitSingle, bytes)
             );
-            IUniswapPermit2(uniswapPermit2).permit(owner, permit, permitSig);
+            IUniswapPermit2(permit2).permit(owner, permit, permitSig);
         }
-        IUniswapPermit2(uniswapPermit2).transferFrom(from, to, uint160(amount), token);
+        IUniswapPermit2(permit2).transferFrom(from, to, uint160(amount), token);
+    }
+
+    function _collectByPermit2SignatureTransfer(
+        address token,
+        address from,
+        address to,
+        uint256 amount,
+        bytes memory data
+    ) private {
+        require(data.length > 0, "TokenCollector: permit2 data cannot be empty");
+        (IUniswapPermit2.PermitTransferFrom memory permit, IUniswapPermit2.SignatureTransferDetails memory detail, address owner, bytes memory permitSig) = abi
+            .decode(data, (IUniswapPermit2.PermitTransferFrom, IUniswapPermit2.SignatureTransferDetails, address, bytes));
+        require(permit.permitted.token == token, "TokenCollector: token address doesn't match");
+        require(owner == from, "TokenCollector: from address doesn't match");
```

Comment:
What's the purpose to have `owner` encoded in data? I think maybe this check can be left to `permitTransferFrom()` of Permit2 by brining `from` address directly to the function.

Reply:
I agree that directly calling permit2 with `from` as owner is a good way, but this would cause client to encode data by our data format instead of permit2's interface.

I would prefer to let clients stick to permit2 native interface, so they don't need to learn additional specs in order to transfer tokens. 

Would love to hear more thoughts from others.

Reply:
But encoding data with our format is inevitable and it could also reduce the length of calldata. The user may use permit2 SDK so having the same format may avoid re-encoding, but the re-encoding can save ~30 bytes in calldata which I think is worthy.

Reply:
I do the gas comparison and here is the result: (only run the success case to rule out noises)

* Current
<img width="1351" alt="before" src="https://user-images.githubusercontent.com/10285880/223356597-5ac4753e-b91f-4802-babf-6dd202278cea.png">

* Custom format
<img width="1352" alt="after" src="https://user-images.githubusercontent.com/10285880/223356701-aa1ecbb0-be71-4747-b3cf-18ef41ae1410.png">

Custom format implementation could refer to https://github.com/consenlabs/tokenlon-contracts/compare/permit2-sig-transfer...permit2-gas-diff

To summarize the changes in custom format version:

1. Gas consumption reduces ~1000.
2. The client still needs to sign permit2 struct (`IUniswapPermit2.PermitTransferFrom`) to create valid permit signature. 
    - There is an implicit rule in the struct that the permitted amount should be equal to requested amount.
    - Client has to extract `nonce`, `deadline` from permit2 struct, and combine them with permit signature, to encode the data to us.

We can discuss about which version would you guys prefer. After reaching a consensus, I will go for that direction.

cc. @charlesjhongc @doublespending @NIC619

Reply:
I still put gas consumption at the higher priority so I think only required values should be brought as parameters.

Reply:
Refactor in [db5b60c](https://github.com/consenlabs/tokenlon-contracts/pull/173/commits/db5b60c26d174d72f9f3608c95f84a35fb6f1d33), also refactor allowance transfer in [367767b](https://github.com/consenlabs/tokenlon-contracts/pull/173/commits/367767b43cd1e2d4d4a8b617ddfa1830667b66b5)

Reply:
I prefer to bring values as parameters, for the input parameters of `PermitSingle` are fixed and need not to let them free. Also, we should make sure that the encoded `data` and the inputs of `_collectByPermit2AllownaceTransfer` are match.